netgear/netgear.go

```diff
@@ -0,0 +1,146 @@
+package netgear
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const timeout = 10 * time.Second
+
+type Netgear struct {
+	sessionToken string
+	client       *http.Client
+
+	hostName string
+	username string
+	password string
+}
+
+func NewNetgear(hostName, username, password string) *Netgear {
+	return &Netgear{
+		client:   &http.Client{Timeout: timeout},
+		hostName: strings.TrimSuffix(hostName, "/") + "/api/v1",
+		username: username,
+		password: password,
+	}
+}
+
+func (n *Netgear) Login() error {
+	loginPayload := map[string]interface{}{
+		"user": map[string]string{
+			"name":     n.username,
+			"password": n.password,
+		},
+	}
+
+	payloadBytes, err := json.Marshal(loginPayload)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(
+		http.MethodPost,
+		fmt.Sprintf("%s/login", n.hostName),
+		bytes.NewBuffer(payloadBytes),
+	)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := n.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	var result map[string]any
+	if err := json.Unmarshal(body, &result); err != nil {
+		return fmt.Errorf("failed to parse login response JSON: %w", err)
+	}
+
+	userRaw, ok := result["user"]
+	if !ok {
+		return fmt.Errorf("login response: missing 'user' field; body: %s", string(body))
```

Comment:
I would advise against including the actually response body into an error message. For one, you cannot know how big the body is - potentially ending up with a huuuuge error. Furthermore, it may contain sensitive information.